Clarify context interaction for trace module (#1063)
open-telemetry/opentelemetry-specification: specification/trace/api.md

the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
returned `Tracer`s and actively update their configuration if it changes.

## Context Interaction

This section defines all operations within the Tracing API that interact with the
[`Context`](../context/context.md).

The API MUST provide the following functionality to interact with a `Context`
instance:

- Extract the `Span` from a `Context` instance
- Insert the `Span` to a `Context` instance

The functionality listed above is necessary because API users SHOULD NOT have
access to the [Context Key](../context/context.md#create-a-key) used by the Tracing API implementation.

If the language has support for implicitly propagated `Context` (see
[here](../context/context.md#optional-global-operations)), the API SHOULD also provide
the following functionality:

> - Get the currently active span
> - Set the currently active span

Reviewer: ```suggestion
- Get the currently active span from the implicit context. This is equivalent to getting the implicit context, then extracting the `Span` from the context.
- Set the currently active span to the implicit context. This is equivalent to getting the implicit context, then inserting the `Span` to the context.
```

All the above functionalities operate solely on the context API, and they MAY be
exposed as static methods on the trace module, as static methods on a class
inside the trace module (it MAY be named `TracingContextUtilities`), or on the

> [`Tracer`](#tracer) class. This functionality SHOULD be fully implemented in the

Reviewer 2: Breaking news - there seems to be a lot of support for removing the mention of `Tracer` here. What do you think?

https://github.com/open-telemetry/opentelemetry-java/issues/1807

Reviewer 3: The spec here does not say that this should be on the tracer, and if it is, it says it should be static.

Reviewer 2: Ah static is written twice, but not a third time. So with this wording, static does not apply to Tracer I think and I didn't read it that way. If it's also static it makes sense to me.

Reviewer 4: What one language decided to do is not breaking news :) In the Ruby implementation they want to work directly with `Tracer` (cc @mwear). The options are given above, and the language is free to implement them either way they want. 

@bogdandrutu to just clarify, we still allow this to exist as instance methods in `Tracer`, correct?

Reviewer 3: I think we should discourage an instance method on tracer, as this has been known to cause confusion and has caused people to call for allowing an unnamed tracer just to access the current span.

Reviewer 5: > I think we should discourage an instance method on tracer

I agree.

Author: I think the current text does exactly that, and in edge cases where that cannot be avoided, it allows to be an instance method.

Reviewer 6: https://github.com/open-telemetry/opentelemetry-ruby/pull/425#issuecomment-710096680

TL;DR: I agree, including for Ruby.

the `TracerProvider`, or the `TracerProvider` could maintain a registry of all
returned `Tracer`s and actively update their configuration if it changes.

## Context Interaction

> This section defines all operations within trace module that interact with the

Reviewer 2: I think this is the first time I see the word "trace module", is tracing like before more consistent?

Reviewer 7: ```suggestion
This section defines all operations within the Tracing API that interact with the
```

[`Context`](../context/context.md).

The API MUST provide the following functionality to interact with a `Context`
instance:

- Extract the `Span` from a `Context` instance
- Insert the `Span` to a `Context` instance

The functionality listed above is necessary because API users SHOULD NOT have

> access to the [Context's Key](../context/context.md#create-a-key).

Reviewer 7: There can be many keys in a context. What about:
```suggestion
access to the [Context Key](../context/context.md#create-a-key) used by the Tracing API implementation.
```

> If the languages has support for implicitly propagated `Context`, see

Reviewer 2: ```suggestion
If the language has support for implicitly propagated `Context`, see
```

> If the language has support for implicitly propagated `Context`, see
> [here](../context/context.md#optional-global-operations). The API SHOULD provide
> the following functionality:

Reviewer 7: ```suggestion
If the language has support for implicitly propagated `Context` (see
[here](../context/context.md#optional-global-operations)), the API SHOULD also provide
the following functionality:
```

Otherwise it's not clear that the condition applies to that sentence as well and not only to when to "see here".

- Get the currently active span
- Set the currently active span

> All the above functionalities operate solely on the context API, and they MAY be

Reviewer 4: Use "either" here? Otherwise it may be interpreted as languages having the freedom to expose this one or more of the mentioned locations - which is exactly what we don't want :)

Reviewer 2: I don't understand the difference between "all" and "either". So I'm trying to interpret the point about not having the freedom to expose in one or more locations - are you saying we don't want e.g., `TracerProvider` to give access to the current span, and `Tracer` to delegate to that as a shortcut? Or `static Span.current` delegate to something like `getGlobalTracerProvider().getCurrentSpan()`? Shortcuts can make users delighted, and I don't think the spec should forbid them.

Author: Are you suggesting to encourage duplicate code and functionality in the API? I feel that is very dangerous for us, and will require a lot of work to maintain backwards compatibility, and will make progress harder.

I would be against us allowing duplicate functionality in the API (avoid confusion for the user), but I am ok to split this into 2 parts (as described by the specs) if that makes things simpler to use.

Reviewer 2: I'm not familiar enough with every programming language to know whether allowing or denying multiple ways to do the same thing is a pattern or antipattern. In Java, shortcuts are common I think FWIW. 

But anyways I would want the language SIGs to focus on making the API that they feel is easiest to use for users and the spec must not block that. It can provide suggestions for sure but it must not block languages from creating a proper, but of course functional (as defined by spec) API.

This is instrumentation, so it's very different from defining remote APIs with a single protobuf and accepting it's at times annoying API where reduction of idioms is more accepted.

Author: Sure, let's get an example where you say (Java):
* https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/TracingContextUtils.java#L43
* https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/Tracer.java#L76

Can you explain how this improve usability? I am worried more that this confuses users:
* which one do I use?
* The tracer version seems to be implementation specific, so that should be used?

I am all about having shortcuts for "getCurrentSpan" instead of asking users to do "getSpan(getCurrentContext())" which is improvement in usability. But having the same shortcut "getCurrentSpan" in multiple places I cannot see how that can improve usability and not confuse users.

I know that it can be said that Java did a poor job (which I agree), but Java was following what was documented in the Spec, which was 2 different shortcuts to achieve the same thing.

Reviewer 2: I agree that a confusing shortcut isn't helpful :) `TracingContextUtils` is a special, and somewhat unnecessary concept so it's not really a helpful API IMO. But if a user wants a current `Span`, having a shortcut on the `Span` itself seems very easy to use, even if it's a shorcut for getting the current tracer provider and its current span. The locations are somewhat different, but it's intuitive I think. Maybe it isn't though - I just wouldn't have the spec define this strictly. I don't see a problem with the current language, it uses `MAY`. But I hope people are on the same page that whether a shortcut is a good idea or not can be a language-issue and not a spec issue.

Author: I agree that sometimes shortcuts are useful, my advice start small, provide a way, if feedback suggest not enough extend. it is simpler to add more ways than removing things, so if we add too many shortcuts and we confuse users it is harder to fix.

> exposed as static methode on the trace module, as static methods on a class

Reviewer 2: ```suggestion
exposed as static methods on the trace module, as static methods on a class
```

inside the trace module (it MAY be named `TracingContextUtilities`), or on the

> [`Tracer`](#tracer) class.

Reviewer 3: I think we are again missing notes on overridablility here. Probably we want this to be fully implemented in the API again and not overridable. If it's in the tracer or tracerprovider, that would be different, then I would expect it to be overridable.

Reviewer 4: Would a note on users NOT being encouraged to override this would be enough? I feel that for some languages it makes more sense to have these methods in the `Tracer`, and being overridable is only an unintended side effect.

Author: Not sure, that's why I did not have any opinion about this. Should I document that it is "not encouraged to override this"?

Reviewer 4: > Should I document that it is "not encouraged to override this"?

Sounds good to me. Opinion on this @Oberon00 ?

cc @mwear

Reviewer 2: For context interoperability, I think we will override methods related to context similar to here

https://github.com/open-telemetry/opentelemetry-java/blob/master/context/src/otelAsGrpcTest/java/io/opentelemetry/context/GrpcContextStorageProvider.java

I wouldn't add a point about not encouraging overriding since for interop of implicit context with other libraries it can be very important.

Reviewer 4: @anuraaga Overriding the low-level `Context` storage has always existed and is totally fine (Python also does it, IIRC).

What we don't want however, for simplicity purposes, is to have users override either the more-public `TracingContextUtils` or `Tracer`/`Baggage` active handling members.

Reviewer 3: I think by default if something is described in the API, I would assume:

* It MUST be overridable
* The API-only implementation MUST be no-op.

For this functionality, I don't think we want any of these, so I suggest to use the same wording we already have for the propagated span and SpanContext creation:

> This functionality MUST be fully implemented in the API, and SHOULD NOT be overridable.

Reviewer 4: @Oberon00 Agreed, that's the idea.

Reviewer 3: I generally agree with this PR but I feel the sentence

> This functionality MUST be fully implemented in the API, and SHOULD NOT be overridable.

(especially the first part) is really important to not muddle the distinction between API and SDK and make every implementer guess what is to be implemented where.

Thus I'm requesting changes here. But please dismiss my request changes once you fixed that and count my review as an approval (I probably won't be here later today).

Author: @Oberon00 added a less strict sentence, there may be case where this cannot be achieved. Take a look and let me know if you are ok with that.

Reviewer 3: If we expect cases where this is not possible, why doesn't the same apply to propagated span creation and spancontext creation?

Author: > Ok, maybe because we have a dependency on a context module only from the SDK implementation, right? Could be the case in JS @dyladan ?

That's what I have in mind, if Context cannot be fully implemented and needs an implementation, then we need to have a way to use that implementation to implement this functionality, so we cannot implement this. Unless we pass the "implementation" to these methods/funcs.

- [Create a new `Span`](#span-creation) (see the section on `Span`)

> The `Tracer` MAY provide functions to:

Reviewer 8: I would prefer that we keep the option for these methods to be on a tracer instance. I made a prototype for Ruby (https://github.com/open-telemetry/opentelemetry-ruby/pull/425) that is implemented this way. Instead of implementing them as class methods, I introduced the notion of a default tracer. When a user doesn't already have a handle on a tracer, they can use the default tracer for span operations such as these. These allows for a clean API as all span operations live on a tracer instance.

Author: See line 157, what is wrong with that? It says the functionality can leave on the tracer class, but avoids duplicating of the same functionality

Reviewer 8: The wording led me to be believe that the methods should be class (static) methods and not instance methods on the class. As long as it allows for them to be instance methods on the tracer, I'm 👍.
